### ISSUE TITLE
XLA inserts a per-iteration copy of the entire scan carry when a passthrough lax.cond is in the loop body

### ISSUE THREAD

**Issue by oliverdiamond** [jax-ml/jax#38145]
Hi! Thank you for maintaining this awesome project! I was looking for some assistance with the following issue (I used an llm to formalize the findings, apologies if this is too verbose). 

### Description

# (0.7.2 → 0.8.2): XLA inserts a per-iteration copy of the entire `scan` carry when a passthrough `lax.cond` is in the loop body

## Summary

A `lax.scan` that carries some parameters and, each step, uses a `lax.cond` to
*either* update them (true branch) *or* return them unchanged (false branch /
passthrough) compiles to very different HLO between jaxlib 0.7.2 and 0.8.2:

- **0.7.2**: the while-body forwards the carry directly — **0 copies/iter**.
- **0.8.2**: the while-body defensively **copies every carried buffer, every
  iteration** — for a 25k-step scan this is ~tens of GB of extra memory traffic
  and ~75k extra copy-kernel launches.

On an L40S this is a **2.6× slowdown** (251 ms → 651 ms) on a tiny benchmark,
with everything except the jaxlib version held constant. Note that I also found the issues 
exists in 0.9.1 as well, I just only have wheels for 0.7.2, 0.8.2 and 0.9.1 on my cluster. 

It is `cond`-specific: with the predicate constant (so the cond folds away) both
versions are identical. The copies only appear once a real conditional survives
in the loop body.

This pattern is common in RL training loops (e.g. a DQN that does a gradient
update every N env steps via `lax.cond(is_learn_time, learn, noop)` inside a
`scan` over env steps), where the carried state is the whole train state —
making the per-step copy very expensive in practice.

## Repro

Self-contained, `jax` + `jax.numpy` only:

```python
import jax, jax.numpy as jnp

MATMUL_DIM, ACTION_DIM = 512, 18
SCAN_STEPS, BACKWARD_EVERY, LR = 25_000, 10_000, 6.25e-5

r1, r2, r3 = jax.random.split(jax.random.PRNGKey(42), 3)
params = {
    "W1": jax.random.normal(r1, (MATMUL_DIM, MATMUL_DIM)) * 0.01,
    "W2": jax.random.normal(r2, (MATMUL_DIM, MATMUL_DIM)) * 0.01,
    "W3": jax.random.normal(r3, (MATMUL_DIM, ACTION_DIM)) * 0.01,
}

def forward(p, x):
    x = jax.nn.relu(x @ p["W1"])
    x = jax.nn.relu(x @ p["W2"])
    return x @ p["W3"]

grad_fn = jax.value_and_grad(lambda p, x: forward(p, x).sum())

def step(carry, _):
    params, i = carry
    def fwd_bwd(_):
        _, g = grad_fn(params, jnp.ones((1, MATMUL_DIM)))
        return jax.tree_util.tree_map(lambda p, g: p - LR * g, params, g)
    def passthrough(_):
        return params
    new_params = jax.lax.cond(
        i % BACKWARD_EVERY == BACKWARD_EVERY - 1, fwd_bwd, passthrough, None)
    return (new_params, i + 1), None

@jax.jit
def run(params):
    (final, _), _ = jax.lax.scan(step, (params, jnp.int32(0)), None, length=SCAN_STEPS)
    return final

print(jax.jit(run).lower(params).compile().as_text())  # inspect the while body
```

> Note: `BACKWARD_EVERY` must be > 1. At `== 1` the predicate is constant-true,
> XLA folds the cond away, and there is no conditional (and no regression).

## Observed vs expected HLO (optimized), while-body ROOT tuple

**0.7.2 (expected)** — carry forwarded directly:

```
ROOT %tuple.23 = (...) tuple(%loop_add_fusion,
    %get-tuple-element.94, %get-tuple-element.95, %get-tuple-element.96)
```

**0.8.2 (regressed)** — full copy of each carried buffer inserted first:

```
%copy.15 = f32[512,512]{1,0} copy(%get-tuple-element.94)
%copy.16 = f32[512,512]{1,0} copy(%get-tuple-element.95)
%copy.17 = f32[512,18]{1,0}  copy(%get-tuple-element.96)
ROOT %tuple.23 = (...) tuple(%loop_add_fusion,
    %copy.15, %copy.16, %copy.17)
```

The conditional itself is preserved (real branching) in **both** versions; the
only difference is the unconditional per-iteration copy of the carry in 0.8.2.

## A/B (only jaxlib version differs)

| | jax 0.7.2 | jax 0.8.2 |
|---|---|---|
| per-iteration copies in while-body | **0** | **3** |
| mean wall time (25k steps) | **251 ms** | **651 ms** |
| GPU / driver / runtime CUDA | L40S / 595.71.05 / 13.2 | same |
| XLA compile toolkit (`--xla_gpu_cuda_data_dir`) | cudacore 12.9.1 | same |
| numpy / python / host | 2.4.2 / 3.11.4 / same node | same |

Wall-time delta (~400 ms) is consistent with ~75k extra copy-kernel launches
(3 copies × 25k iters), i.e. launch-overhead bound rather than bandwidth bound.

## Environment

`jax.print_environment_info()` on the **regressed** build:

```
jax:    0.8.2
jaxlib: 0.8.2
numpy:  2.4.2
python: 3.11.4 (main, Nov  4 2023, 03:43:13) [GCC 12.3.1 20230526]
device info: NVIDIA L40S-1, 1 local devices
process_count: 1
platform: uname_result(system='Linux', node='rack09-15', release='6.8.0-117-generic', version='#117-Ubuntu SMP PREEMPT_DYNAMIC Tue May  5 19:26:24 UTC 2026', machine='x86_64')
```

`jax.print_environment_info()` on the **good** build:

```
jax:    0.7.2
jaxlib: 0.7.2
numpy:  2.4.2
python: 3.11.4 (main, Nov  4 2023, 03:43:13) [GCC 12.3.1 20230526]
device info: NVIDIA L40S-1, 1 local devices
process_count: 1
platform: uname_result(system='Linux', node='rack09-15', release='6.8.0-117-generic', version='#117-Ubuntu SMP PREEMPT_DYNAMIC Tue May  5 19:26:24 UTC 2026', machine='x86_64')
```

Both run with `XLA_FLAGS=--xla_gpu_cuda_data_dir=/cvmfs/soft.computecanada.ca/easybuild/software/2023/x86-64-v3/Core/cudacore/12.9.1`.
GPU: NVIDIA L40S, driver 595.71.05 (max CUDA 13.2 per `nvidia-smi`). Packages are
`jax`/`jaxlib`/`jax_cuda12_plugin`/`jax_cuda12_pjrt`, all `+computecanada` at the
matching version (0.8.2 or 0.7.2).

## Caveats

- These are `+computecanada` builds, not stock PyPI wheels — but a friend was able to 
reproduce the issue with the official `jax==0.8.2` / `0.7.2` wheels, and the only
  package that changes between the two environments is jaxlib (and its matching
  CUDA plugin/pjrt at the same version).

**Comment by Gattocrucco** [comment]
(Note: I hit this bug as well.)

**Comment by Gattocrucco** [comment]
I opened an issue for the underlying cause of this problem, #39449. The workaround in your MWE is to replace

```python
jax.lax.cond(i % BACKWARD_EVERY == BACKWARD_EVERY - 1, fwd_bwd, passthrough, None)
```

with 

```python
jax.lax.switch((i % BACKWARD_EVERY == BACKWARD_EVERY - 1).astype(jnp.int32), [passthrough, fwd_bwd], None)
```

**Comment by SakshamKapoor2911** [comment]
This was fixed in openxla/xla PR #46314 (https://github.com/openxla/xla/pull/46314) and merged to main in commits bb290bd (https://github.com/openxla/xla/commit/bb290bdc0d11330392cf23673cf2817baf915429) / 124548f (https://github.com/openxla/xla/commit/124548f7f8336e598a5c233e9cf8d58d13a25133). It was also exported to tensorflow/tensorflow via commit ec59954 (https://github.com/tensorflow/tensorflow/commit/ec5995434e2c08bbbe2094318bbe039d5d640b7b).

The root cause was asymmetric copy elision in kConditional; HLO's buffer-sharing logic used a direction-dependent flag that only elided copies for branch 0. The fix replaces it with proper escape analysis, so a passthrough branch at any position forwards its operand without copies.

GitHub's cross-repo auto-close didn't trigger because the commit was pushed by the copybara bot, which lacks write permissions on jax-ml/jax. Could a maintainer verify whether this issue is resolved and close it?
